### PR TITLE
Refactoring Wallet / Cryptography

### PR DESCRIPTION
**Pull request by pierluca**
At this point, please only review commit 670e32482832c13c3d463f5ed7b6de206e935d34
It still needs testing, but it should clarify the general direction.

### CODE REVIEW

File: fe1-web/navigation/index.ts [hunk 1]
```diff
@@ -0,0 +1,3 @@
+/**
+ * The navigation module contains the elements
```
Comment (nkcr) [inline fe1-web/navigation/index.ts:2]: ...the navigation elements ?

**Review by nkcr — APPROVED** [review]

**Comment by pierluca** [comment]
Please do not checkout this branch locally - I'll make some changes to the git history to clean it up.

**Comment by pierluca** [comment]
@nkcr sorry about this ridiculously big PR 😞 Cleaning up the wallet ended up forcing me to change things everywhere.
A future refactoring shouldn't have this effect, as things should be much more de-coupled now.

Regarding the wallet refactoring, you want to have a look at (in this order, from lowest-level to highest-level):
* ./store/reducers/WalletReducer.ts 
* ./store/stores/WalletStore.ts (and its dependencies, i.e. ./platform/ )
* ./model/objects/wallet/*.ts (and its associated tests)

I'm not sure if the wallet code is in the right place, it seems weird to put it in `./model/objects` considering that while it's semantically an object, it's ultimately implemented as a bunch of functions. I'm almost tempted to put it in its own `./wallet/` root-level folder. Let me know what you think.

Other changes include:

* Implementing the roll call messages creation logic
* Moving `__tests__` in their respective folders
* Simplifying the mocking & testing logic for readability - this is just a first step and MUCH more is needed.

**Comment by pierluca** [comment]
Reviewed Sonar's feedback:
* Tests still need to be refactored, I won't fix those as part of this PR
* I'm minded to merge in a single object OpenRollCall and ReopenRollCall to address the massive code duplication.

**Review by nkcr — COMMENTED** [review]
Minor comments. Haven't looked at every details but overall that feels much better than before!

Yes, `wallet` could be moved somewhere else than in `./model`. However if there is a `./wallet` folder, then I would expect other features to be there, like with an `./election` folder. Would it make sense then to have a `./feature` or `./domain` folder containing folders of feature-specific elements ? (i.ie `./domain/wallet`, `./domain/election`, ...)

File: fe1-web/model/objects/wallet/Bip39Path.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ */
+function segmentsFromHash(hash: Hash): string {
+  const buffer = hash.toBuffer();
+
+  let segments: string[] = [];
+  for (let i = 0; i < buffer.length; i += 3) {
+    const seg = buffer.slice(i, i + 3).join('');
+    segments = segments.concat(seg);
+  }
+  return segments.join(HARDENED_SYMBOL + PATH_SEPARATOR) + HARDENED_SYMBOL;
```
[the remark this answers is not in the document]
Comment (pierluca) [reply fe1-web/model/objects/wallet/Bip39Path.ts:28]: easier to understand, but wrong 😅  
my code:
`seg1 seg2 seg3 HARD SEP seg4 seg5 seg6 SEP `

your suggestion:
`seg1 HARD seg2 HARD seg3 HARD SEP seg4 HARD seg5 HARD seg6 HARD `

**Comment by pierluca** [comment]
I would fully support a `./feature` folder, that's a very different (if typical) project structure though, where reducers, GUI and business logic are grouped by feature. Or we can do a hybrid.
More fundamentally, how can we keep the project structure consistency if doing a transition? I don't want students to put things in all sorts of (wrong) places :)

**Review by nkcr — APPROVED** [review]

**Comment by nkcr** [comment]
> More fundamentally, how can we keep the project structure consistency if doing a transition? I don't want students to put things in all sorts of (wrong) places :)

For that decision I rely on your knowledge of the code base and experience with students. We should also take that discussion in live. In the mean time for me it's OK to merge like that.

**Pull request merged**